README.md
# quote-it

[![build status](https://img.shields.io/travis/imcuttle/quote-it/master.svg?style=flat-square)](https://travis-ci.org/imcuttle/quote-it)
[![Test coverage](https://img.shields.io/codecov/c/github/imcuttle/quote-it.svg?style=flat-square)](https://codecov.io/github/imcuttle/quote-it?branch=master)
[![NPM version](https://img.shields.io/npm/v/quote-it.svg?style=flat-square)](https://www.npmjs.com/package/quote-it)
[![NPM Downloads](https://img.shields.io/npm/dm/quote-it.svg?style=flat-square&maxAge=43200)](https://www.npmjs.com/package/quote-it)

Safe quoting a given string without adding duplicate quotes

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### quote

[index.js:46-57](https://github.com/imcuttle/quote-it/blob/5c1b0291e0cc2742427b102f31b68d1ca0cf1f87/index.js#L46-L57 'Source code on GitHub')

Uses `quoteChar` to wrap string.

#### Parameters

- `string` {string}
- `quoteChar` {string}

#### Examples

```javascript
import quote from 'quote-it'
quote('abc', "'") === "'abc'"
```

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**

### single

[index.js:68-70](https://github.com/imcuttle/quote-it/blob/5c1b0291e0cc2742427b102f31b68d1ca0cf1f87/index.js#L68-L70 'Source code on GitHub')

Uses single quote to wrap string.

#### Parameters

- `string`

#### Examples

```javascript
import { single } from 'quote-it'
single("a'bc") === "'a'bc'"
```

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**

### double

[index.js:78-80](https://github.com/imcuttle/quote-it/blob/5c1b0291e0cc2742427b102f31b68d1ca0cf1f87/index.js#L78-L80 'Source code on GitHub')

Uses double quote to wrap string.

#### Parameters

- `string`

Returns **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)**

## License

MIT


CHANGELOG.md
<a name="1.0.1"></a>

## 1.0.1 (2018-08-20)


__tests__/main.test.js
/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
import { double, single, default as quote } from '../index'

it('should quote-it double passed', function() {
  expect(double('sbac')).toBe('"sbac"')
  expect(double('sb/ac')).toBe('"sb/ac"')
  expect(double('sb\\ac')).toBe('"sb\\\\ac"')
  expect(double('sb\\ac')).toBe(JSON.stringify('sb\\ac'))
  expect(double('sbsac')).toBe('"sbsac"')
  expect(double('sb"a\'c')).toBe('"sb\\"a\'c"')
})

it('should spec character', function() {
  expect(quote('hello[wo\u200brld', '"')).toBe('"hello[wo\u200brld"')
  expect(quote('hello[wo\u200brld', '"')).toBe(JSON.stringify('hello[wo\u200brld'))
  expect(quote('hello[wo\x00crld', '"')).toBe('"hello[wo\\u0000crld"')
  expect(quote('hello[wo\x00crld', '"')).toBe(JSON.stringify('hello[wo\x00crld'))
  expect(quote('hello[wo\x1fcrld', '"')).toBe(JSON.stringify('hello[wo\x1fcrld'))
  expect(quote('hello[wo\x1fcrld', '"')).toBe('"hello[wo\\u001fcrld"')
})

it('should quote-it single passed', function() {
  expect(single("'sbac")).toBe("'\\'sbac'")
  expect(single("sb'a'c\"")).toBe("'sb\\'a\\'c\"'")
})

it('should quote-it other character passed', function() {
  expect(quote('hello%world', '%')).toBe('%hello\\%world%')
  expect(quote('hello-world', '-')).toBe('-hello\\-world-')
  expect(quote('hello$world', '$')).toBe('$hello\\$world$')
  expect(quote('[]hel\\lo[]world[]', '[]')).toBe('[][]hel\\lo[]world[][]')
  expect(quote('hello world', 'w')).toBe('whello \\worldw')
})


index.js
// Forked from https://github.com/substack/jsonify/blob/master/lib/stringify.js

function getEscapable(char) {
  char = char === '-' ? '\\-' : char
  return new RegExp(`[\\\\${char}\x00-\x1f]`, 'g')
}

const commonMeta = {
  // table of character substitutions
  '\b': '\\b',
  '\t': '\\t',
  '\n': '\\n',
  '\f': '\\f',
  '\r': '\\r',
  '\\': '\\\\'
}

function _quote(string, char = '"', reg = escapable, meta) {
  // If the string contains no control characters, no quote characters, and no
  // backslash characters, then we can safely slap some quotes around it.
  // Otherwise we must also replace the offending characters with safe escape
  // sequences.
  reg.lastIndex = 0
  return reg.test(string)
    ? char +
        string.replace(reg, function(a) {
          var c = meta[a]
          // control characters [\x00-\x1f] converts to \uxxxx
          return typeof c === 'string' ? c : '\\u' + ('0000' + a.charCodeAt(0).toString(16)).slice(-4)
          // : a
        }) +
        char
    : char + string + char
}

/**
 * Uses `quoteChar` to wrap string.
 * @public
 * @param string {string}
 * @param quoteChar {string}
 * @return {string}
 * @example
 * import quote from 'quote-it'
 * quote("abc", "'") === "'abc'"
 */
export default function quote(string, quoteChar = '"') {
  if (quoteChar.length > 1 && process.env.NODE_ENV !== 'production') {
    console.error('quote: `quoteChar` is recommended as single character, but ' + JSON.stringify(quoteChar) + '.')
  }

  return _quote(
    string,
    quoteChar,
    getEscapable(quoteChar),
    Object.assign({}, commonMeta, { [quoteChar]: `\\${quoteChar}` })
  )
}

/**
 * Uses single quote to wrap string.
 * @public
 * @param string
 * @return {string}
 * @example
 * import { single } from 'quote-it'
 * single("a'bc") === "'a\'bc'"
 */
export function single(string) {
  return quote(string, "'")
}

/**
 * Uses double quote to wrap string.
 * @public
 * @param string
 * @return {string}
 */
export function double(string) {
  return quote(string, '"')
}


index.es5.js
'use strict'

exports.__esModule = true
exports.default = quote
exports.single = single
exports.double = double
// Forked from https://github.com/substack/jsonify/blob/master/lib/stringify.js

function getEscapable(char) {
  char = char === '-' ? '\\-' : char
  return new RegExp('[\\\\' + char + '\0-\x1F]', 'g')
}

var commonMeta = {
  // table of character substitutions
  '\b': '\\b',
  '\t': '\\t',
  '\n': '\\n',
  '\f': '\\f',
  '\r': '\\r',
  '\\': '\\\\'
}

function _quote(string) {
  var char = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '"'
  var reg = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : escapable
  var meta = arguments[3]

  // If the string contains no control characters, no quote characters, and no
  // backslash characters, then we can safely slap some quotes around it.
  // Otherwise we must also replace the offending characters with safe escape
  // sequences.
  reg.lastIndex = 0
  return reg.test(string)
    ? char +
        string.replace(reg, function(a) {
          var c = meta[a]
          // control characters [\x00-\x1f] converts to \uxxxx
          return typeof c === 'string' ? c : '\\u' + ('0000' + a.charCodeAt(0).toString(16)).slice(-4)
          // : a
        }) +
        char
    : char + string + char
}

/**
 * Uses `quoteChar` to wrap string.
 * @public
 * @param string {string}
 * @param quoteChar {string}
 * @return {string}
 * @example
 * import quote from 'quote-it'
 * quote("abc", "'") === "'abc'"
 */
function quote(string) {
  var _Object$assign

  var quoteChar = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '"'

  if (quoteChar.length > 1 && process.env.NODE_ENV !== 'production') {
    console.error('quote: `quoteChar` is recommended as single character, but ' + JSON.stringify(quoteChar) + '.')
  }

  return _quote(
    string,
    quoteChar,
    getEscapable(quoteChar),
    Object.assign(
      {},
      commonMeta,
      ((_Object$assign = {}), (_Object$assign[quoteChar] = '\\' + quoteChar), _Object$assign)
    )
  )
}

/**
 * Uses single quote to wrap string.
 * @public
 * @param string
 * @return {string}
 * @example
 * import { single } from 'quote-it'
 * single("a'bc") === "'a\'bc'"
 */
function single(string) {
  return quote(string, "'")
}

/**
 * Uses double quote to wrap string.
 * @public
 * @param string
 * @return {string}
 */
function double(string) {
  return quote(string, '"')
}
